// tailwindcss-3d/utils/lodash-transformers.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    axesModifier: function() {
        return axesModifier;
    },
    nameModifier: function() {
        return nameModifier;
    },
    signModifier: function() {
        return signModifier;
    },
    addDurationWithGravity: function() {
        return addDurationWithGravity;
    }
});
var _lodash = /*#__PURE__*/ _interopRequireDefault(require("lodash"));
function _defineProperty(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _objectSpread(target) {
    for(var i = 1; i < arguments.length; i++){
        var source = arguments[i] != null ? arguments[i] : {};
        var ownKeys = Object.keys(source);
        if (typeof Object.getOwnPropertySymbols === "function") {
            ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function(sym) {
                return Object.getOwnPropertyDescriptor(source, sym).enumerable;
            }));
        }
        ownKeys.forEach(function(key) {
            _defineProperty(target, key, source[key]);
        });
    }
    return target;
}
var axesModifier = function(axes, silentModifier) {
    var safeAxes = _lodash.default.isUndefined(axes) ? [
        "x",
        "y",
        "z"
    ] : _lodash.default.flatten([
        axes
    ]);
    var safeSilentModifier = _lodash.default.isUndefined(silentModifier) ? "1" : silentModifier;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return function(result, value, modifier) {
        _lodash.default.forEach(safeAxes, function(axis) {
            var property = String(modifier) === String(safeSilentModifier) ? axis : "".concat(axis, "-").concat(modifier);
            result[property] = _lodash.default.isPlainObject(value) ? _objectSpread({
                axis: axis,
                originalModifier: modifier
            }, value) : {
                axis: axis,
                originalModifier: modifier,
                value: value
            };
        });
    };
};
var nameModifier = function(name, silentModifier) {
    var safeName = _lodash.default.isString(name) ? name : "";
    var safeSilentModifier = _lodash.default.isString(silentModifier) ? silentModifier : safeName === "" // If there is no name, don't silence the modifier
     ? "" : "1";
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return function(result, value, modifier) {
        var property = String(modifier) === safeSilentModifier ? safeName : safeName === "" ? modifier : "".concat(safeName, "-").concat(modifier);
        result[property] = _lodash.default.isPlainObject(value) ? _objectSpread({
            name: name,
            originalModifier: modifier
        }, value) : {
            name: name,
            originalModifier: modifier,
            value: value
        };
    };
};
var signModifier = function(signs) {
    var safeSigns = _lodash.default.isUndefined(signs) ? [
        "",
        "-"
    ] : _lodash.default.flatten([
        signs
    ]);
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return function(result, value, modifier) {
        _lodash.default.forEach(safeSigns, function(sign) {
            result["".concat(sign).concat(modifier)] = _lodash.default.isPlainObject(value) ? _objectSpread({
                sign: sign,
                originalModifier: modifier
            }, value) : {
                sign: sign,
                originalModifier: modifier,
                value: value
            };
        });
    };
};
var addDurationWithGravity = function() {
    return(// eslint-disable-next-line @typescript-eslint/no-explicit-any
    function(result, value, modifier) {
        var _value_originalModifier;
        var duration = Math.pow(Number((_value_originalModifier = value.originalModifier) !== null && _value_originalModifier !== void 0 ? _value_originalModifier : modifier), 0.25).toFixed(2);
        result[modifier] = _lodash.default.isPlainObject(value) ? _objectSpread({
            duration: duration,
            originalModifier: modifier
        }, value) : {
            duration: duration,
            originalModifier: modifier,
            value: value
        };
    });
};
