// tailwindcss-3d/css-utilities/perspective-origin.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "perspectiveOrigin", {
    enumerable: true,
    get: function() {
        return perspectiveOrigin;
    }
});
var _defaultTheme = /*#__PURE__*/ _interopRequireDefault(require("tailwindcss/defaultTheme"));
function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError("Cannot call a class as a function");
    }
}
function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
var PerspectiveOrigin = function PerspectiveOrigin() {
    "use strict";
    _classCallCheck(this, PerspectiveOrigin);
    this.defaultTheme = _defaultTheme.default.transformOrigin;
    this.utilities = function(param) {
        var matchUtilities = param.matchUtilities, theme = param.theme;
        matchUtilities({
            "perspective-origin": function(value) {
                return {
                    "perspective-origin": value
                };
            }
        }, {
            values: theme("perspectiveOrigin")
        });
    };
};
var perspectiveOrigin = new PerspectiveOrigin();
