// tailwindcss-3d/css-utilities/perspective.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "perspective", {
    enumerable: true,
    get: function() {
        return perspective;
    }
});
function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError("Cannot call a class as a function");
    }
}
var Perspective = function Perspective() {
    "use strict";
    _classCallCheck(this, Perspective);
    this.defaultTheme = {
        none: "none",
        250: "250px",
        500: "500px",
        750: "750px",
        1000: "1000px"
    };
    this.utilities = function(param) {
        var matchUtilities = param.matchUtilities, theme = param.theme;
        matchUtilities({
            perspective: function(value) {
                return {
                    "@defaults transform": {},
                    "--tw-perspective": value,
                    perspective: "var(--tw-perspective)"
                };
            }
        }, {
            values: theme("perspective"),
            supportsNegativeValues: true
        });
    };
};
var perspective = new Perspective();
