// tailwindcss-3d/css-utilities/transform-core.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "transformCore", {
    enumerable: true,
    get: function() {
        return transformCore;
    }
});
var _translate = require("./translate");
var _scale = require("./scale");
var _transform = require("./transform");
function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError("Cannot call a class as a function");
    }
}
var TransformCore = function TransformCore() {
    "use strict";
    _classCallCheck(this, TransformCore);
    this.utilities = function(param) {
        var addUtilities = param.addUtilities;
        addUtilities({
            ".transform": {
                transform: _transform.transform.normaliseFunctionValues({
                    dimension: "2d"
                })
            },
            ".transform-cpu": {
                translate: _translate.translate.normaliseFunctionValues({
                    dimension: "2d"
                }),
                scale: _scale.scale.normaliseFunctionValues({
                    dimension: "2d"
                }),
                transform: _transform.transform.normaliseFunctionValues({
                    dimension: "2d"
                })
            },
            ".transform-gpu": {
                translate: _translate.translate.normaliseFunctionValues(),
                scale: _scale.scale.normaliseFunctionValues(),
                transform: _transform.transform.normaliseFunctionValues()
            },
            ".transform-none": {
                transform: "none"
            }
        });
    };
};
var transformCore = new TransformCore();
