// tailwindcss-3d/css-utilities/transform-box.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "transformBox", {
    enumerable: true,
    get: function() {
        return transformBox;
    }
});
function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError("Cannot call a class as a function");
    }
}
var TransformBox = function TransformBox() {
    "use strict";
    _classCallCheck(this, TransformBox);
    this.defaultTheme = {
        content: "content-box",
        border: "border-box",
        fill: "fill-box",
        stroke: "stroke-box",
        view: "view-box"
    };
    this.utilities = function(param) {
        var matchUtilities = param.matchUtilities, theme = param.theme;
        matchUtilities({
            "transform-box": function(value) {
                return {
                    "transform-box": value
                };
            }
        }, {
            values: theme("transformBox")
        });
    };
};
var transformBox = new TransformBox();
